provider/datagen/src/crabbake.rs

```diff
@@ -0,0 +1,339 @@
+    mod_files: Mutex<HashMap<PathBuf, Vec<String>>>,
+    // List of dependencies used by baking.
+    dependencies: CrateEnv,
+}
+
+impl ConstExporter {
+    pub fn new(mod_directory: PathBuf, pretty: bool, insert_feature_gates: bool) -> Self {
+        let _ = std::fs::remove_dir_all(&mod_directory);
+        Self {
+            mod_directory,
+            pretty,
+            insert_feature_gates,
+            data: Default::default(),
+            per_key_data: Default::default(),
+            mod_files: Default::default(),
+            dependencies: Default::default(),
+        }
+    }
+
+    fn write_to_file<P: AsRef<std::path::Path>>(
+        &self,
+        path: P,
+        data: TokenStream,
+    ) -> Result<(), DataError> {
+        let path = self.mod_directory.join(path).with_extension("rs");
+        std::fs::create_dir_all(&path.parent().unwrap())?;
+
+        {
+            let mut file = line_ending_file::BufWriterWithLineEndingFix::new(File::create(&path)?);
+            writeln!(file, "// GENERATED MODULE. DO NOT EDIT")?;
+            writeln!(file, "{}", data)?;
+        }
+
+        if self.pretty {
+            std::process::Command::new("rustfmt")
+                .arg(path)
+                // The default, "auto", is meant to detect the existing line endings and preserve them.
+                // However, this seems to be broken and generates Unix line endings on Windows, which
+                // introduces Git diffs when regenerating.
+                .args(&["--config", "newline_style=native"])
+                .spawn()
+                .unwrap()
+                .wait()?;
+        }
+        Ok(())
+    }
+}
+
+impl DataExporter<CrabbakeMarker> for ConstExporter {
+    fn put_payload(
+        &self,
+        key: ResourceKey,
+        options: ResourceOptions,
+        payload: DataPayload<CrabbakeMarker>,
+    ) -> Result<(), DataError> {
+        let (payload, marker_type) = payload.tokenize(&self.dependencies);
+        let mut map = self.data.lock().expect("poison");
+        if !map.contains_key(&key) {
+            map.insert(key, (marker_type.to_string(), LiteMap::new()));
+        }
+        map.get_mut(&key)
+            .unwrap()
+            .1
+            .insert(options, payload.to_string());
+        Ok(())
+    }
+
+    fn flush(&self, key: ResourceKey) -> Result<(), DataError> {
+        if let Some((marker, raw)) = self.data.lock().expect("poison").remove(&key) {
+            let marker = syn::parse_str::<syn::Path>(&marker).unwrap();
+
+            let mut sorted = raw.into_tuple_vec();
+            // Sort by payload bake so we can deduplicate.
+            sorted.sort_by(|(_, a), (_, b)| a.cmp(b));
+
+            let mut statics = Vec::new();
+            let mut all_options = Vec::new();
+
+            for (payload_bake_string, group) in &sorted
+                .into_iter()
+                .group_by(|(_, payload_bake_string)| payload_bake_string.clone())
```

Comment:
Thought: I'm a little surprised that it seems `Itertools::group_by` doesn't let the closure return something that borrows